Sources/HTTP/SimpleSocket/SimpleServerSocket.swift

```diff
@@ -0,0 +1,152 @@
+// This source file is part of the Swift.org Server APIs open source project
+//
+// Copyright (c) 2017 Swift Server API project authors
+// Licensed under Apache License v2.0 with Runtime Library Exception
+//
+// See http://swift.org/LICENSE.txt for license information
+//
+
+import Foundation
+import Dispatch
+
+internal class SimpleServerSocket {
+    internal var socketfd: Int32 = -1
+    
+    internal var listeningPort: Int32 = -1
+    
+    internal private(set) var isListening = false
+    
+    internal private(set) var isConnected = false
+
+    internal func socketRead(into readBuffer: inout UnsafeMutablePointer<Int8>, maxLength:Int) throws -> Int {
+        //let readBuffer: UnsafeMutablePointer<CChar> = UnsafeMutablePointer<CChar>.allocate(capacity: maxLength)
+        readBuffer.initialize(to: 0x0)
+        let read = recv(self.socketfd, readBuffer, maxLength, Int32(0))
+//        if read > 0 {
+//            data.append(readBuffer, length: read)
+//        }
+        return read
+    }
+    
+    @discardableResult internal func socketWrite(from buffer: UnsafeRawPointer, bufSize: Int) throws -> Int {
+        return send(self.socketfd, buffer, Int(bufSize), Int32(0))
+    }
+    
+    internal func shutdownAndClose() {
+        //print("Shutting down socket \(self.socketfd)")
+        if self.isListening || self.isConnected {
+            //print("Shutting down socket")
+            _ = shutdown(self.socketfd, Int32(SHUT_RDWR))
```

Comment:
Is the `_ = ` necessary?

Reply:
It generates a warning if you ignore the result. It's not marked @discarable in Glibc (or Darwin or both)

Reply:
Is it safe to discard it then?

Reply:
Should throw if the return != 0

Reply:
The reason I had been avoiding `throw`ing if `shutdown(2)` fails is because it shouldn't be an error if `shutdownAndClose` gets called on an already-closed-and-invalidated socket. Does that seem reasonable?